fedimint-client/src/sm/executor.rs

```diff
@@ -0,0 +1,1161 @@
+        self.valid_module_ids.insert(context.module_instance_id());
+
+        if self
+            .module_contexts
+            .insert(context.module_instance_id(), context)
+            .is_some()
+        {
+            panic!("Tried to add two modules with the same instance id!");
+        }
+    }
+
+    /// Allow executor to build state machines associated with the module id,
+    /// for which the module itself might not be available yet (otherwise it
+    /// would be registered with `[Self::with_module_dyn]`).
+    pub fn with_valid_module_id(&mut self, module_id: ModuleInstanceId) {
+        self.valid_module_ids.insert(module_id);
+    }
+
+    /// Build [`Executor`] and spawn background task in `tasks` executing active
+    /// state machines. The supplied database `db` must support isolation, so
+    /// cannot be an isolated DB instance itself.
+    pub fn build(self, db: Database, notifier: Notifier, client_task_group: TaskGroup) -> Executor {
+        let (sm_update_tx, sm_update_rx) = tokio::sync::mpsc::unbounded_channel();
+
+        let inner = Arc::new(ExecutorInner {
+            db,
+            state: std::sync::RwLock::new(ExecutorState::Unstarted { sm_update_rx }),
+            module_contexts: self.module_contexts,
+            valid_module_ids: self.valid_module_ids,
+            notifier,
+            sm_update_tx,
+            client_task_group,
+        });
+
+        debug!(
+            target: LOG_CLIENT_REACTOR,
+            instances = ?inner.module_contexts.keys().copied().collect::<Vec<_>>(),
+            "Initialized state machine executor with module instances"
+        );
+        Executor { inner }
+    }
+}
+#[derive(Debug)]
+pub struct ActiveOperationStateKeyPrefix {
+    pub operation_id: OperationId,
+}
+
+impl Encodable for ActiveOperationStateKeyPrefix {
+    fn consensus_encode<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        self.operation_id.consensus_encode(writer)
+    }
+}
+
+impl ::fedimint_core::db::DatabaseLookup for ActiveOperationStateKeyPrefix {
+    type Record = ActiveStateKeyDb;
+}
+
+#[derive(Debug)]
+pub(crate) struct ActiveModuleOperationStateKeyPrefix {
+    pub operation_id: OperationId,
+    pub module_instance: ModuleInstanceId,
+}
+
+impl Encodable for ActiveModuleOperationStateKeyPrefix {
+    fn consensus_encode<W: Write>(&self, writer: &mut W) -> Result<(), Error> {
+        self.operation_id.consensus_encode(writer)?;
+        self.module_instance.consensus_encode(writer)?;
+        Ok(())
+    }
+}
+
+impl ::fedimint_core::db::DatabaseLookup for ActiveModuleOperationStateKeyPrefix {
+    type Record = ActiveStateKeyDb;
+}
+
+#[derive(Debug)]
+pub struct ActiveStateKeyPrefix;
+
+impl Encodable for ActiveStateKeyPrefix {
+    fn consensus_encode<W: Write>(&self, _writer: &mut W) -> Result<(), Error> {
+        Ok(())
+    }
+}
+
+#[derive(Encodable, Decodable, Debug)]
+pub struct ActiveStateKeyDb(pub fedimint_client_module::sm::executor::ActiveStateKey);
```

Comment:
Since `ActiveStateKey` and `InactiveStateKey` are available to the module code, they can't be used as as keys in the client's database. A newtype wrappers are used to decouple these.